File: src/AutoContext.Worker.Web/src/hosting/pipe-framing.ts
import type { Readable, Writable } from 'node:stream';

/**
 * Maximum payload size accepted by {@link readMessage}. Caps allocation
 * when a corrupted or hostile header arrives; tasks exchanged on this
 * pipe are small JSON envelopes well below this limit.
 */
export const MAX_MESSAGE_BYTES = 64 * 1024 * 1024;

/**
 * Length-prefixed binary framing helpers for the worker pipe protocol:
 * 4-byte little-endian payload length followed by that many UTF-8 JSON
 * bytes. Matches the C# `PipeFraming` helper in
 * `AutoContext.Worker.Shared` bit-for-bit.
 */

/**
 * Reads one length-prefixed message from {@link stream}. Returns
 * `null` when the connection is closed before a full header is
 * received. Throws on header overflow or when cancellation is signaled
 * mid-read.
 */
export async function readMessage(stream: Readable, signal?: AbortSignal): Promise<Buffer | null> {
    const header = await readExact(stream, 4, signal);

    if (header === null) {
        return null;
    }

    const length = header.readInt32LE(0);

    if (length <= 0) {
        return Buffer.alloc(0);
    }

    if (length > MAX_MESSAGE_BYTES) {
        throw new Error(
            `Pipe message length ${length} exceeds the maximum of ${MAX_MESSAGE_BYTES} bytes.`,
        );
    }

    return readExact(stream, length, signal);
}

/**
 * Writes {@link payload} as one length-prefixed message on
 * {@link stream}.
 */
export async function writeMessage(
    stream: Writable,
    payload: Buffer,
    signal?: AbortSignal,
): Promise<void> {
    throwIfAborted(signal);

    const message = Buffer.allocUnsafe(4 + payload.length);
    message.writeInt32LE(payload.length, 0);
    payload.copy(message, 4);

    await new Promise<void>((resolve, reject) => {
        let abortListener: (() => void) | undefined;

        const cleanup = (): void => {
            if (abortListener !== undefined) {
                signal?.removeEventListener('abort', abortListener);
            }
        };

        if (signal !== undefined) {
            abortListener = (): void => {
                cleanup();
                reject(signalAbortError(signal));
            };
            signal.addEventListener('abort', abortListener, { once: true });
        }

        stream.write(message, (err) => {
            cleanup();
            if (err !== null && err !== undefined) {
                reject(err);
                return;
            }
            resolve();
        });
    });
}

function readExact(
    stream: Readable,
    byteCount: number,
    signal: AbortSignal | undefined,
): Promise<Buffer | null> {
    return new Promise((resolve, reject) => {
        if (signal?.aborted === true) {
            reject(signalAbortError(signal));
            return;
        }

        if (byteCount === 0) {
            resolve(Buffer.alloc(0));
            return;
        }

        let settled = false;

        const settle = (action: () => void): void => {
            if (settled) {
                return;
            }
            settled = true;
            stream.removeListener('readable', onReadable);
            stream.removeListener('end', onEnd);
            stream.removeListener('error', onError);
            if (signal !== undefined) {
                signal.removeEventListener('abort', onAbort);
            }
            action();
        };

        const tryRead = (): void => {
            if (settled) {
                return;
            }
            // `read(n)` returns exactly `n` bytes when available, or
            // null otherwise — avoiding the unshift/re-entry pitfalls
            // of the flowing 'data' mode. At EOF with fewer than `n`
            // bytes buffered, Node returns the partial remainder; we
            // treat that as a clean "stream ended" outcome.
            const chunk = stream.read(byteCount) as Buffer | null;
            if (chunk === null) {
                return;
            }
            if (chunk.length < byteCount) {
                settle(() => resolve(null));
                return;
            }
            settle(() => resolve(chunk));
        };

        const onReadable = (): void => {
            tryRead();
        };

        const onEnd = (): void => {
            settle(() => resolve(null));
        };

        const onError = (err: Error): void => {
            settle(() => reject(err));
        };

        const onAbort = (): void => {
            settle(() => reject(signalAbortError(signal!)));
        };

        stream.on('readable', onReadable);
        stream.on('end', onEnd);
        stream.on('error', onError);
        if (signal !== undefined) {
            signal.addEventListener('abort', onAbort, { once: true });
        }

        // Data may already be buffered (common for back-to-back reads
        // on the same stream); try once before relying on events.
        tryRead();
    });
}

function throwIfAborted(signal: AbortSignal | undefined): void {
    if (signal?.aborted === true) {
        throw signalAbortError(signal);
    }
}

function signalAbortError(signal: AbortSignal): Error {
    const reason: unknown = signal.reason;
    if (reason instanceof Error) {
        return reason;
    }
    const err = new Error('The operation was aborted.');
    err.name = 'AbortError';
    return err;
}

File: src/AutoContext.Worker.Shared/Hosting/PipeFraming.cs
namespace AutoContext.Worker.Hosting;

using System.Buffers.Binary;

/// <summary>
/// Length-prefixed binary framing helpers for the worker pipe protocol:
/// 4-byte little-endian payload length followed by that many UTF-8 JSON bytes.
/// </summary>
public static class PipeFraming
{
    /// <summary>
    /// Maximum payload size accepted by <see cref="ReadMessageAsync"/>.
    /// Caps allocation when a corrupted or hostile header arrives; tasks
    /// exchanged on this pipe are small JSON envelopes well below this limit.
    /// </summary>
    public const int MaxMessageBytes = 64 * 1024 * 1024; // 64 MiB

    /// <summary>
    /// Reads one length-prefixed message from <paramref name="stream"/>.
    /// Returns <see langword="null"/> when the connection is closed before
    /// a full header is received.
    /// </summary>
    public static async Task<byte[]?> ReadMessageAsync(Stream stream, CancellationToken ct)
    {
        ArgumentNullException.ThrowIfNull(stream);

        var header = new byte[4];
        var headerRead = await ReadExactAsync(stream, header, ct).ConfigureAwait(false);

        if (!headerRead)
        {
            return null;
        }

        var length = BinaryPrimitives.ReadInt32LittleEndian(header);

        if (length < 0)
        {
            throw new InvalidDataException(
                $"Pipe message length {length} is negative; header is corrupt.");
        }

        if (length == 0)
        {
            return [];
        }

        if (length > MaxMessageBytes)
        {
            throw new InvalidDataException(
                $"Pipe message length {length} exceeds the maximum of {MaxMessageBytes} bytes.");
        }

        var payload = new byte[length];
        var payloadRead = await ReadExactAsync(stream, payload, ct).ConfigureAwait(false);

        return payloadRead ? payload : null;
    }

    /// <summary>
    /// Writes <paramref name="payload"/> as one length-prefixed message.
    /// </summary>
    public static async Task WriteMessageAsync(Stream stream, byte[] payload, CancellationToken ct)
    {
        ArgumentNullException.ThrowIfNull(stream);
        ArgumentNullException.ThrowIfNull(payload);

        var message = new byte[4 + payload.Length];
        BinaryPrimitives.WriteInt32LittleEndian(message, payload.Length);
        payload.CopyTo(message.AsSpan(4));

        await stream.WriteAsync(message, ct).ConfigureAwait(false);
        await stream.FlushAsync(ct).ConfigureAwait(false);
    }

    private static async Task<bool> ReadExactAsync(Stream stream, byte[] buffer, CancellationToken ct)
    {
        var offset = 0;

        while (offset < buffer.Length)
        {
            var read = await stream.ReadAsync(
                buffer.AsMemory(offset, buffer.Length - offset), ct).ConfigureAwait(false);

            if (read == 0)
            {
                return false;
            }

            offset += read;
        }

        return true;
    }
}
